### PR TITLE
ITF ADR update

### PR DESCRIPTION
**Pull request by Kukovec**
- [ ] Tests added for any new code
- [ ] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] [Entries added to `./unreleased/`][changelog format] for any new functionality

First half of #2258 

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

File: docs/src/adr/015adr-trace.md [hunk 1]
```diff
@@ -230,11 +230,11 @@ as `CONSTANTS` in TLA+. For example, the field may look like:
   "params": [ "Missionaries", "Cannibals" ]
 ```
 
-The field `vars` is an array of names that must be set in every state.
+The field `vars` is an object, the keys of which are names that must be set in every state, and the values are string representations of their types (as defined in [ADR002]).
 For example, the field may look like:
 
 ```js
-  "vars": [ "bank_of_boat", "who_is_on_bank" ]
+  "vars": { "bank_of_boat": "Str", "who_is_on_bank": "Str -> Set(PERSON)" }
```
Comment (thpani) [inline docs/src/adr/015adr-trace.md:237]: Could "vars" carry more information in the future?
A more future-proof / backwards-compatible encoding would be

```suggestion
  "vars": [ { "name": "bank_of_boat", "type": "Str" },
            { "name": "who_is_on_bank", "type": "Str -> Set(PERSON)" } ]
```

**Review by thpani — COMMENTED** [review]
I haven't really followed the discussion on this, so I'll leave the review to @shonfeder.

One question on the chosen encoding though:

**Comment by thpani** [comment]
afaict @konnov authored this ADR, should we wait for his review next week?

**Comment by Kukovec** [comment]
> afaict @konnov authored this ADR, should we wait for his review next week?

We can, but then I'll be blocked until he comes back + we have consensus.

**Comment by shonfeder** [comment]
I think Igor's point about using the meta to maintain backwards compat makes sense.

Or, if we are going to break it, given we have known users, we should be planning a versioned format release, so that might call for a new ADR (or just an issue) explaining the versioning scheme, and then implementing the breaking changes on top of that.

But imo, just putting the variable type data in the meta seems most expedient.

**Comment by shonfeder** [comment]
Oh, I see. Maybe your suggestion for a new field for this is the good middle path then?

**Review by konnov — COMMENTED** [review]
As we briefly discussed online, I would prefer to have the type data in the `#meta` object. The reasons are:

 - The types syntax and the type system may change in the future. Moreover, with the current approach there is no issue of having traces that use different type systems.
 - The types are optional, e.g., if we produce an ITF trace with TLC.
 - In principle, the type information is redundant. Most of the time, it could be inferred from the right-hand sides of the assignments. However, types cannot be uniquely inferred for empty sets and empty functions.

Additionally, the current format is as minimal as possible. All additional info goes into the meta object. This gives us a bit of resiliency w.r.t. the format changes. All tools should support the absolute minimum. If they need additional data, it is optionally provided in the meta object.

**Pull request merged**